yourbench/pipeline/upload_ingest_to_hub.py

```diff
@@ -0,0 +1,236 @@
+            Additional metadata, such as file size or arbitrary user-defined fields.
+    """
+
+    document_id: str
+    document_text: str
+    document_filename: str
+    document_metadata: dict[str, Any] = field(default_factory=dict)
+
+
+def run(config: dict[str, Any]) -> None:
+    """
+    Primary function to execute the 'upload_ingest_to_hub' stage.
+
+    This function aggregates markdown documents from a given source directory
+    (configured in `pipeline.upload_ingest_to_hub.source_documents_dir`) into a
+    Hugging Face Dataset, which is then saved locally or pushed to the Hub.
+
+    Args:
+        config (dict[str, Any]):
+            The overall pipeline configuration dictionary. Relevant keys:
+
+            - config["pipeline"]["upload_ingest_to_hub"]["run"] (bool):
+                Whether to run this stage.
+            - config["pipeline"]["upload_ingest_to_hub"]["source_documents_dir"] (str):
+                Directory path for the ingested markdown files.
+            - config["hf_configuration"]["token"] (str, optional):
+                Hugging Face token for authentication if uploading a private dataset.
+            - config["hf_configuration"]["private"] (bool):
+                Whether to keep the dataset private on the Hub (defaults to True).
+            - config["hf_configuration"]["global_dataset_name"] (str):
+                Base dataset name on Hugging Face (can be overridden).
+            - config["pipeline"]["upload_ingest_to_hub"]["output_dataset_name"] (str, optional):
+                The name of the dataset to save to/push to on the Hugging Face Hub.
+            - config["pipeline"]["upload_ingest_to_hub"]["output_subset"] (str, optional):
+                Subset name for partial saving (default is this stage name).
+
+    Raises:
+        ValueError:
+            If `source_documents_dir` is missing in the config, indicating incomplete config.
+    """
+    stage_name = "upload_ingest_to_hub"
+    stage_cfg = config.get("pipeline", {}).get(stage_name, {})
+
+    # Check if this stage is turned off in config
+    if not stage_cfg.get("run", False):
+        logger.info(f"Stage '{stage_name}' is disabled. Skipping.")
+        return
+
+    source_dir: Optional[str] = stage_cfg.get("source_documents_dir")
+    if not source_dir:
+        error_msg = (
+            f"Missing required field 'source_documents_dir' in pipeline.{stage_name}. "
+            f"Cannot proceed with uploading ingested documents."
+        )
+        logger.error(error_msg)
+        raise ValueError(error_msg)
+
+    # Determine the final dataset name and subset from config
+    hf_cfg = config.get("hf_configuration", {})
+    output_dataset_name = stage_cfg.get("output_dataset_name", hf_cfg.get("global_dataset_name"))
+    output_subset = stage_cfg.get("output_subset", stage_name)
+
+    # Show key info about Hugging Face Hub config
+    hf_token: Optional[str] = hf_cfg.get("token", None)
```

Comment:
we will want to grab it from the environment or hf_hub in the future instead of forcing the user to have it as string in the config